scalanet/src/io/iohk/scalanet/peergroup/kademlia/KPeerGroup.scala

```diff
@@ -0,0 +1,161 @@
+package io.iohk.scalanet.peergroup.kademlia
+
+import java.net.{InetAddress, InetSocketAddress}
+
+import io.iohk.scalanet.peergroup.PeerGroup.ServerEvent.ChannelCreated
+import io.iohk.scalanet.peergroup.PeerGroup.{ChannelSetupException, HandshakeException, ServerEvent}
+import io.iohk.scalanet.peergroup.kademlia.KPeerGroup.{ChannelImpl, NodeRecord, UnderlyingChannel}
+import io.iohk.scalanet.peergroup.{Channel, InetMultiAddress, PeerGroup}
+import monix.eval.Task
+import monix.execution.Scheduler
+import monix.reactive.Observable
+import monix.reactive.subjects.Subject
+import org.slf4j.{Logger, LoggerFactory}
+import scodec.bits.BitVector
+
+// PeerGroup using node records for addressing.
+// These node records are derived from Ethereum node records (https://eips.ethereum.org/EIPS/eip-778)
+class KPeerGroup[M](
+    val kRouter: KRouter[NodeRecord],
+    val server: Subject[ServerEvent[BitVector, M], ServerEvent[BitVector, M]],
+    underlyingPeerGroup: PeerGroup[InetMultiAddress, Either[NodeRecord, M]]
+)(implicit scheduler: Scheduler)
+    extends PeerGroup[BitVector, M] {
+
+  private val log = LoggerFactory.getLogger(getClass)
+
+  override def processAddress: BitVector = kRouter.config.nodeId
+
+  override def initialize(): Task[Unit] = {
+    // The protocol defined here requires that a peer
+    // connects then sends it node record as the first message
+    // (it could send earlier messages but they will be ignored)
```

Comment:
where is this procedure described?
According to the paper, page 7, last paragraph of section 2.3:
To join the network, a node u must have contact to an already participating node w. Then:
- u inserts w in its corresponding bucket.
- u then performs a node lookup for its own ID.
- u then refreshes all its buckets with the triples it gets from the FIND_NODE queries.
- The queries made by u also may update other nodes buckets.

Reply:
As discussed, it's entirely separate from kademlia.